core/src/main/java/org/infinispan/protostream/impl/TagWriterImpl.java

```diff
@@ -794,6 +819,51 @@ void writeBytes(ByteBuffer value) throws IOException {
          }
       }
 
+      @Override
+      void writeUTF8Field(int number, String s) throws IOException {
+         if (!(out instanceof LazyByteArrayOutputStream os)) {
+            super.writeUTF8Field(number, s);
+            return;
+         }
+
+         int strlen = s.length();
+         int tag = WireType.makeTag(number, WireType.WIRETYPE_LENGTH_DELIMITED);
+         int varIntTagLen = varIntSize(tag);
+         int varIntBytesLen = varIntSize(strlen);
+         int startPos = os.getPosition();
+
+         // First optimize for 1 - 127 case
+         os.ensureCapacity(startPos + varIntTagLen + varIntBytesLen + strlen);
+         byte[] buf = os.getRawBuffer();
+         writeVarInt32Direct(tag, startPos, buf);
+         // Note this will be overwritten if not all 1 - 127 characters below
+         writeVarInt32Direct(strlen, startPos + varIntTagLen, buf);
+
+         int localPos = startPos + varIntTagLen + varIntBytesLen;
+
+         int c;
+         int i;
+         for (i = 0; i < strlen; i++) {
+            c = s.charAt(i);
+            if (c > 127) break;
+
+            buf[localPos++] = (byte) c;
+         }
+
+         os.setPosition(localPos);
```

Comment:
I believe this only has to be in the if statement below.

Reply:
We need it in either case. If it's a multi-byte string, then we call `#ensureCapacity` which utilises the position to copy all existing bytes to the array created with the new capacity.